lib/SILGen/SILGenExpr.cpp

```diff
@@ -2351,9 +2350,11 @@ static bool mayLieAboutNonOptionalReturn(Expr *expr) {
     return mayLieAboutNonOptionalReturn(declRef->getDecl());
   }
 
-  // An application, which we look through to get the function we're calling.
+  // An application that produces a reference type, which we look through to
+  // get the function we're calling.
   if (auto apply = dyn_cast<ApplyExpr>(expr)) {
-    return mayLieAboutNonOptionalReturn(apply->getFn());
+    return apply->getType()->hasReferenceSemantics() &&
```

Comment:
We need this extra check because ApplyExprs can only lie about their return nullability if the result of the application is nullable in C. The existing code recurs into the ApplyExpr's subexpression but loses that context.